BUG: Adds support for array parameter declaration in fortran code

melissawm opened the pull request:
Hello,

This PR is supposed to add support for array parameter declaration in fortran code. Previously, this resulted in an error when trying to use array parameter values to define other variables in the same code (see bug #11612)

I also added some tests, including one which is supposed to fail right now because this fix does not include the special case of array parameters with 0-based indexing in fortran. 

Closes #11612 
Closes #8730

melissawm committed:
BUG, ENH: Adds support for parameter arrays in fortran code

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -579,8 +579,8 @@ def readfortrancode(ffile, dowithline=show, istop=1):
-intrisicpattern = re.compile(
-    beforethisafter % ('', 'intrisic', 'intrisic', '.*'), re.I), 'intrisic'
+intrinsicpattern = re.compile(
+    beforethisafter % ('', 'intrinsic', 'intrinsic', '.*'), re.I), 'intrinsic'
```
It might make sense to do all these typo fixes in a standalone PR.

pearu reviewed (changes requested):
Thanks for working on this! I agree that the PR could be split into several smaller ones each fixing one issue at the time. Otherwise, PR-s may grow with no limit.

mattip commented on numpy/f2py/crackfortran.py:
```diff
@@ -2470,6 +2493,7 @@ def get_parameters(vars, global_params={}):
 
             if isdouble(vars[n]):
                 tt = list(v)
+                print(real16pattern.finditer(v))
```
still not done?

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2481,11 +2504,30 @@ def get_parameters(vars, global_params={}):
+                    except Exception as msg:
+                        v_eval.append(item)
+                        outmess('get_parameters: got "%s" on %s\n' % (msg, repr(v)))
+                params[n] = v_eval.copy()
```
```suggestion
                params[n] = v_eval
```
No need to make a copy of a temporary you never use again

eric-wieser commented on numpy/f2py/tests/test_parameter.py:
```diff
@@ -114,3 +116,18 @@ def test_constant_sum(self):
+        assert_equal(z, 18.0)
+
+    @pytest.mark.xfail
+    # No support for zero-based indexed parameter arrays
```
```suggestion
    @pytest.mark.xfail(reason="No support for zero-based indexed parameter arrays")
```

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2481,11 +2504,30 @@ def get_parameters(vars, global_params={}):
+                        v_eval.append(eval(item, g_params, params))
+                    except Exception as msg:
+                        v_eval.append(item)
+                        outmess('get_parameters: got "%s" on %s\n' % (msg, repr(v)))
```
```suggestion
                    try:
                        item = eval(item, g_params, params)
                    except Exception as msg:
                        outmess('get_parameters: got "%s" on %s\n' % (msg, repr(v)))
                    v_eval.append(item)
```

melissawm commented on numpy/f2py/crackfortran.py:
```diff
@@ -2481,11 +2505,30 @@ def get_parameters(vars, global_params={}):
+                v = v.lstrip('(/').rstrip('/)').split(',')
+                # TODO Probably not the most efficient way...
+                v_eval = []
+                for item in v:
```
I will do this in a separate PR if that's ok.

melissawm replied:
Sorry, I ended up doing this all here since I couldn't solve one problem without touching the other. Let me know if you see any problems with the code here

melissawm commented on numpy/f2py/crackfortran.py:
```diff
@@ -2650,11 +2700,28 @@ def analyzevars(block):
+                        dname = d[:d.find("(")]
+                        ddims = d[d.find("(")+1:d.find(")")]
+                        # FIXME this is not robust; only works if the
+                        # parameter array is defined with 1-based indexing
```
This will also come in a separate PR.

melissawm replied:
I also think this is fixed for now

melissawm commented:
Hi all, I just added a few comments to let you know that I'm planning on working on more delicate issues in separate PR's so we can do proper testing and review. The code above works for simple cases. I don't know if this process is the recommended one, let me know what you think. Thanks!

melissawm commented on numpy/f2py/crackfortran.py:
```diff
@@ -2650,11 +2700,28 @@ def analyzevars(block):
+                        ddims = d[d.find("(")+1:d.find(")")]
+                        # FIXME this is not robust; only works if the
+                        # parameter array is defined with 1-based indexing
+                        if int(ddims)-1  < 0:
```
Same as above (will come in a separate PR)

melissawm replied:
I did my best to fix this, hope it works now.

mattip commented:
Just to be clear: you want to merge this as-is with the problematic indexing? If so, could you add an `TODO: fix ...` in the code?

eric-wieser commented:
My understanding was that #15706 was extracted from this, so should be merged first.

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2481,11 +2504,31 @@ def get_parameters(vars, global_params={}):
+            param_is_array = False
+            for a in vars[n]['attrspec']:
+                if a[:9] == 'dimension':
+                    param_is_array = True
```
nit: possibly clearer and certainly shorter as:
```suggestion
            param_is_array = any(a[:9] == 'dimension' for a in vars[n]['attrspec'])
```

melissawm commented:
Sorry, this was a bit messy but I think I have a better version now. I also added two tests, one which is meant to record the fact that for now we can't parse multidimensional parameter arrays, since this would imply executing fortran code before the extension module is built (e.g. the intrinsic `reshape` function).

The code changed quite a bit because I tried incorporating all suggestions while also improving my own understanding of how crackfortran works. Thanks for reviewing and for the patience in waiting for the fix!

melissawm commented on numpy/f2py/crackfortran.py:
```diff
@@ -2650,11 +2700,28 @@ def analyzevars(block):
+                        # the dimension for this variable depends on a
+                        # previously defined (array) parameter
+                        dname = d[:d.find("(")]
+                        ddims = d[d.find("(")+1:d.find(")")]
```
I think this is fixed

melissawm commented:
@pearu can you take a look when you have the time? Thanks!

eric-wieser commented on numpy/f2py/capi_maps.py:
```diff
@@ -22,7 +22,7 @@
 from . import cb_rules
 
-# The environment provided by auxfuncs.py is needed for some calls to eval.
+# The eviroment provided by auxfuncs.py is needed for some calls to eval.
```
Looks like your rebase has gone wrong, this reverts the spelling mistakes you fixed!

eric-wieser replied:
This comment still applies

eric-wieser replied:
```suggestion
# The environment provided by auxfuncs.py is needed for some calls to eval.
```

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2906,84 @@ def analyzevars(block):
+    # 2) allow the parameter array to be accessed in python as a dictionary with fortran
+    #    indices as keys
+    # We are choosing 2 for now.
+    if any(a[:9] == 'dimension' for a in varsn['attrspec']):
+        # This is an array parameter.
+        # First, we parse the dimension information
+        a = [item[9:] for item in varsn['attrspec'] if item[:9] == 'dimension'][0]
```
```suggestion
    dim_attrspecs = [item[9:] for item in varsn['attrspec'] if item[:9] == 'dimension']
    if len(dim_attrspecs) > 1:
        raise ValueError('dimension may appear at most once")
    elif dim_attrspecs:
        # This is an array parameter.
        # First, we parse the dimension information
        a = dim_attrspecs[0]
```

melissawm replied:
The code above selects the first 9 characters of `varsn['attrspec']` to see if dimension is mentioned and then selects the dimension expression in `a` (something like a shape). Your suggestion selects the shape in `dim_attrspecs` and raises a `ValueError` if the shape has length > 1, which is not supposed to be a problem (for example if the declaration is `dimension(2)` then `dim_attrspecs == "(2)"` which has length 3. I guess what you were suggesting is that we don't have dimension specified twice in the declaration of this parameter?

eric-wieser replied:
I don't believe your comment describes either what your original code does or what my suggestion does.

> guess what you were suggesting is that we don't have dimension specified twice in the declaration of this parameter?

Yes, this was my intent

melissawm replied:
Sorry I had missed something. I see it now. Thanks for that!

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2906,84 @@ def analyzevars(block):
+        # This is an array parameter.
+        # First, we parse the dimension information
+        a = [item[9:] for item in varsn['attrspec'] if item[:9] == 'dimension'][0]
+        dimrange = a.lstrip('(').rstrip(')')
```
probably safer as
```suggestion
        a = a.strip()  # does fortran care about whitespace?
        if a[0] != "(" or a[-1] != ")":
            raise ValueError(...)
        dimrange = a[1:-1]
```
Otherwise you could strip off unbalanced parentheses

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2906,84 @@ def analyzevars(block):
+    return p
+
+
+def param_parse(d, params):
```
Could you add a docstring with two examples showing what this function does in the recursive and non-recursive case?

eric-wieser commented on numpy/f2py/tests/src/parameter/constant_array.f90:
```diff
@@ -0,0 +1,30 @@
+  implicit none
+  integer, parameter :: dp = selected_real_kind(15)
+  integer, parameter :: intparamarray(2) = (/ 3, 5 /)
+  real(dp), parameter :: doubleparamarray(3) = (/ 3._dp, 4._dp, 6._dp /)
```
Can you make one of these values be non-integral?

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2906,84 @@ def analyzevars(block):
+    #    is often incompatible with the original fortran indexing)
+    # 2) allow the parameter array to be accessed in python as a dictionary with fortran
+    #    indices as keys
+    # We are choosing 2 for now.
```
One option here would be:
```python
class FortranParamArray:
    def __init__(self, indices, values):
        self._indices = indices
        self._values = values
    def __getitem__(self, index):
        return self._values[self._indices.index(index)]
```

eric-wieser replied:
And pass the `range` object in as the `indices` argument.

melissawm replied:
I didn't consider this because the f2py code does not use OOP and I think this might be a part of a larger refactor in the future.

pearu commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2906,84 @@ def analyzevars(block):
+                dimrange = range(int(dimrange[0]), int(dimrange[1])+1)
+        else:
+            print(n)
+            raise Exception('param_eval: multidimensional array parameters not supported: %s\n' % repr(n))
```
I would just report this via `outmess` rather than raising an exception because the corresponding case is often irrelevant for the f2py task.

pearu reviewed (changes requested):
Thanks, @melissawm for developing `param_eval`!
My only concern is that it is too rigorous in reporting unsupported (read: notimplemented) features as failures while f2py may not need these features when creating the wrapper code.

The used naming convention in "crackfortran.py" has an actual purpose: be forgiving on things that do not matter :)

melissawm commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2906,84 @@ def analyzevars(block):
 analyzeargs_re_1 = re.compile(r'\A[a-z]+[\w$]*\Z', re.I)
 
 
+def param_eval(n, varsn, v, g_params, params):
```
Let me know if the tests I added are enough.

melissawm commented:
I think I addressed everything, feedback is appreciated. Thanks once again for the patience and for the time spent reviewing this!

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2909,141 @@ def analyzevars(block):
+                    ' supported: %s\n' % repr(n))
+
+        # Parse parameter value
+        v = v.lstrip('(/').rstrip('/)').split(',')
```
strip is likely not suitable here, it maps `(/(a), b/)` to `a), b`.

melissawm replied:
I updated the code to address this, but I'm not sure if another approach would be preferable.

pearu replied:
```
v = (v[2:-2] if v.startswith('(/') else v).split(',')
```

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2909,141 @@ def analyzevars(block):
+
+      then `d = 2` and we return immediately, with
+
+      `param_parse(d, params) = '2'`
```
Can you write this in the form of a doctest, with `>>>`?

melissawm replied:
Let me know if this is what you had in mind. I could write a proper doctest but that would involve populating the entire `param` dictionary which I don't think is the idea. I tried looking for similar tests in the rest of the code but couldn't find any. Thanks again for your patience on this, I'm still learning! Feel free to keep commenting if you think anything can be improved.

eric-wieser replied:
I've put a suggestion above - but yes, a minimally populated params dictionary is exactly what I'm looking for, since otherwise it's hard to tell what form that argument takes.

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2909,141 @@ def analyzevars(block):
+                    ' supported: %s\n' % repr(n))
+
+        # Parse parameter value
+        v = v.strip('()').strip('/').split(',')
```
This is probably safe, but @pearu's suggestion was better

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2909,142 @@ def analyzevars(block):
+      then `d = pa(1)`; since `pa` is a previously parsed parameter,
+      and `pa(1) = 3`, we call `param_parse` recursively, to obtain
+
+    >>> d = 'pa(1)'
```
Would this be accurate?
```suggestion
    >>> params = dict(pa={1: 3, 2: 5})
    >>> d = 'pa(1)'
```

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2873,6 +2909,143 @@ def analyzevars(block):
+            index = params[ddims]
+        else:
+            # ddims is a scalar. this means we should just parse it.
+            index = int(ddims)
```
The recursion in this function seems a bit jumbled to me. I'd expect it to be more like
```
if has_paren(d):
    name, ind_str = ...
    index = recourse(ind_str)
    return params[name][index]  # var(i)
elif d in params:
    return params[name]  # var
else:
    return int(d). # i
```

melissawm replied:
Reorganized a bit. Let me know if this is better.

melissawm commented:
Hello, folks. Is there anything else you feel can be done in this PR? I've rebased it just to make sure the tests are still passing.

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -148,7 +148,7 @@
 from . import __version__
 
-# The environment provided by auxfuncs.py is needed for some calls to eval.
+# The enviroment provided by auxfuncs.py is needed for some calls to eval.
```
```suggestion
# The environment provided by auxfuncs.py is needed for some calls to eval.
```

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2925,6 +2961,143 @@ def analyzevars(block):
+        ddims = d[d.find("(")+1:d.rfind(")")]
+    else:
+        return str(d)
+    if '(' in ddims:
+        # this dimension expression is an array element
+        if ddims[:ddims.find("(")] in params:
+            # this dimension expression is also a parameter;
+            # parse it recursively
+            index = int(param_parse(ddims, params))
+            return str(params[dname][index])
+    elif ddims in params:
+        # Reached the bottom of the recursion
+        index = params[ddims]
+        return str(params[dname][index])
```
If the `if` is false here, this returns `None`, rather than the string its supposed to. Are you sure you want to do that?

I'll take another look at this later - I think the logic behind this function is still wrong.

A failing test case would be `      `integer, parameter, dimension(pa) :: pb = (/1, 2, 3/)`, where `pa = 3` - this returns `pa`, but should return `3`.

melissawm replied:
Ok, I'll review this again and let you know when it's ready for your review, I don't want to waste your time if I can find that before you take a look at it. Thanks again!

eric-wieser replied:
I think my complaint is the same as it was in this comment: https://github.com/numpy/numpy/pull/15457#discussion_r488167973. The algorithm I suggest there will pass the test case I give above.

eric-wieser commented on numpy/f2py/tests/src/parameter/constant_array.f90:
```diff
@@ -2,14 +2,16 @@
   integer, parameter :: dp = selected_real_kind(15)
+  integer, parameter :: pa = 2
   integer, parameter :: intparamarray(2) = (/ 3, 5 /)
+  integer, parameter, dimension(pa) :: pb = (/ 2, 10 /)
```
Is this legal fortran?
```
integer, parameter, dimension(intparamarray(1)) :: pc = (/ 2, 10, 20 /)
```
I think this case currently fails.

melissawm replied:
I have added this test and it does not fail. In my understanding, as long as the dimension attribute is a constant expression this is legal under the language specification. However if you feel like this should be investigated further I can check other compilers (my own tests are only run using gfortran for the moment).

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2704,11 +2733,15 @@ def analyzevars(block):
+                        # the dimension for this variable depends on a
+                        # previously defined parameter
+                        d = param_parse(d, params)
                     for p in list(params.keys()):
```
I think this `for` loop actually belongs in the final `else` block of `parse_param`

melissawm commented:
I believe the failures are unrelated, but I'm not sure.

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2704,18 +2733,15 @@ def analyzevars(block):
-                            d = m.group('before') + \
-                                str(params[p]) + m.group('after')
-                            m = re_1.match(d)
+                    if ((d in params) or
+                            (d.find("(") != -1 and d[:d.find("(")] in params)):
+                        # the dimension for this variable depends on a
+                        # previously defined parameter
+                        d = param_parse(d, params)
```
Rather than repeating the logic of `param_parse`, I wonder if this should be
```suggestion
                    try:
                        d = param_parse(d, params)
                    except (ValueError, IndexError, KeyError):
                        pass
```

melissawm replied:
Looks better, thanks!

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2925,6 +2951,154 @@ def analyzevars(block):
+                d = m.group('before') + \
+                    str(params[p]) + m.group('after')
+                m = re_1.match(d)
+        return str(int(d))
```
Is there any benefit to converting to `int` and back here?
```suggestion
        return d
```

melissawm replied:
You're right - I think this was a leftover from some previous change.

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2536,10 +2559,10 @@ def get_parameters(vars, global_params={}):
-                outmess('get_parameters: got "%s" on %s\n' % (msg, repr(v)))
+                outmess('get_parameters: got "%s" on %s\n' % (msg, repr(n)))
+
             if isstring(vars[n]) and isinstance(params[n], int):
```
Won't this line now raise `KeyError` because `params[n]` is no longer set by the `except` block?

melissawm replied:
Yes. Following the suggestion from @pearu above I'm not raising another exception, but I've added the unparsed parameter array as `params[n]` so this doesn't cause trouble further down.

melissawm commented:
@pearu can you take a look when you have the time? Thanks!

pearu reviewed (changes requested):
Thanks, @melissawm for implementing support for array parameters and documenting some of the internal f2py functions!
I have a number of nits and comments that will simplify the code.

melissawm commented:
> Thanks, @melissawm for implementing support for array parameters and documenting some of the internal f2py functions!
> I have a number of nits and comments that will simplify the code.

Thank you, @pearu - I think I've addressed all your comments, please let me know if there's anything else I should look at.

eric-wieser commented on numpy/f2py/f90mod_rules.py:
```diff
@@ -23,7 +23,7 @@
 from .crackfortran import undo_rmbadname, undo_rmbadname1
 
-# The environment provided by auxfuncs.py is needed for some calls to eval.
+# The enviroment provided by auxfuncs.py is needed for some calls to eval.
```
```suggestion
# The environment provided by auxfuncs.py is needed for some calls to eval.
```
I think this is still the same old merge conflict

pearu commented:
@melissawm LGTM, could you address the lint errors?

pearu reviewed (approved):
LGTM

melissawm commented:
Sorry for the ping, but I though @eric-wieser might want to take a last look before this is merged? Thanks!

pearu commented:
I suggest merging this PR, say, within 5 days if there is no reviewing activity, as it is blocking the further development of evaluating expressions in array shapes.

cc: @eric-wieser @mattip

pearu commented:
@melissawm , this PR requires a rebase.

pearu commented:
I did the rebase.

@mattip is it ok to merge this PR as it is (provided that CI is green) or do we need to squash the commit history?

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2954,9 +2991,155 @@ def analyzevars(block):
+            dimrange = range(int(lbound), int(ubound)+1)
+    else:
+        outmess(f'param_eval: multidimensional array parameters ({dimspec}) '
+                'not supported\n')
```
I don't understand this line; we print a message claiming not to support it, but the function keeps running anyway. So either this should raise an exception, or we should adjust the message to indicate it's only _partially_ supported.

melissawm replied:
This was done in response to [@pearu's comment](https://github.com/numpy/numpy/pull/15457#discussion_r445838791):

> I would just report this via outmess rather than raising an exception because the corresponding case is often irrelevant for the f2py task.

eric-wieser replied:
I think it's better to raise the exception - if you do that, then an appropriate place up the stack will emit a suitable error message, and you don't have to guess what the rest of the function does after this message is printed.

eric-wieser replied:
It's consistent with what we do on line 3031

pearu replied:
I disagree with the need to raise an exception: the aim of f2py is to create wrapper functions for Fortran routines and this requires only partial interpretation of Fortran code that may or may not be relevant for creating wrapper functions. Most of the Fortran code is irrelevant for creating the wrapper functions and if one raises an exception on every piece of Fortran code that f2py is not able to interpret, as is the given case here, you basically break f2py.

Recall that f2py does not aim at wrapping arbitrary Fortran code to Python without any user intervention, it would not just be possible in general. According to the recommended usage of f2py, one should use signature files to design the UI and functionality of wrapper functions. Only rarely one is able to apply f2py directly to Fortran source codes and get satisfactory wrapper functions. f2py can be used to generate initial versions of signature files that users ought to modify to their actual needs. If f2py would raise an exception on every Fortran codelet that it does not support yet, one would not be able to generate the signature files. Hence, your suggestion to raise an exception here breaks the normal usage of f2py.

A rule of thumb: in f2py crackfortran.py file, one must not raise exceptions on branches that correspond to f2py not supporting a particular feature in Fortran: such branches are often irrelevant, and printing out a message that a branch of unsupported Fortran construct is met, is sufficient to better design the signature file or submit a feature request for f2py to automatically support the particular Fortran constructs.

Raising an exception in the above-described situations is contrary to f2py model and all such proposals must be rejected.

pearu replied:
However, this is an exception to the above-mentioned rule: when the inputs to the f2py reader are a signature file or an f2py-directive, raising an exception on an unsupported Fortran construct is ok.

eric-wieser replied:
> and if one raises an exception on every piece of Fortran code that f2py is not able to interpret, as is the given case here, you basically break f2py.

You only break f2py if the exception propagates all the way to the caller, which it does not here. It's far easier to reason about functions that raise exceptions when their preconditions are not met and they are unable to proceed - the caller is still free to catch the exception and recover, which is exactly what is happening here.

To emphasize that point - with an exception here, we can add a test for this and verify that it fails as we expect. With a message here, I expect it throws an exception anyway, but it's now a nonsense one cause by continuing with invalid preconditions.

I understand that the goal of f2py is to in many places perform a best effort parsing - but that doesn't mean that every helper function should perform a best effort.

pearu replied:
With https://github.com/numpy/numpy/blob/1adc24adcdc1dcc1d52f8eba2065b6f243980783/numpy/f2py/crackfortran.py#L2596-L2599 I agree with @eric-wieser to raise an exception here or elsewhere as long as one keeps in mind the f2py model: f2py should not fail on Fortran constructs that support has not been implemented.

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2735,19 +2771,20 @@ def analyzevars(block):
+                        # previously defined parameter
+                        d = param_parse(d, params)
+                    except (ValueError, IndexError, KeyError):
+                        pass
```
Should this call `outmess` to log the exception?

melissawm replied:
Done, thanks !

mattip commented:
> do we need to squash the commit history

We are inconsistent in squashing history. If the PR is one small commit and then 10 more "Fix typo", then it is a definite win to squash. But here it is a mix of actual history that may be useful in the future. So if @melissawm would like to squash some of the commits that would be nice, but I wouldn't hold up merging for that.

melissawm commented:
Thanks! I'll to a last pass and respond to @eric-wieser 's comments shortly, and squash the whole thing.

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2567,11 +2590,18 @@ def get_parameters(vars, global_params={}):
             except Exception as msg:
                 params[n] = v
-                outmess('get_parameters: got "%s" on %s\n' % (msg, repr(v)))
+                outmess('get_parameters: got "%s" on %s\n' % (msg, repr(n)))
```
If we change `param_eval` to raise an exception this line will still show the error message and set `params[n] = v`

eric-wieser commented on numpy/f2py/crackfortran.py:
```diff
@@ -2954,9 +2994,155 @@ def analyzevars(block):
+            ubound = param_parse(dimrange[1], params)
+            dimrange = range(int(lbound), int(ubound)+1)
+    else:
+        outmess(f'param_eval: multidimensional array parameters ({dimspec}) '
```
`dimspec` already contains parentheses here, so you probably don't want to add an extra set.

eric-wieser commented on numpy/f2py/tests/src/parameter/constant_array.f90:
```diff
@@ -0,0 +1,45 @@
+  integer, parameter :: pa = 2
+  integer, parameter :: intparamarray(2) = (/ 3, 5 /)
+  integer, parameter, dimension(pa) :: pb = (/ 2, 10 /)
+  integer, parameter, dimension(intparamarray(1)) :: pc = (/ 2, 10, 20 /)
```
Is it legal to have these in a different order like this?
```suggestion
  integer, dimension(intparamarray(1)), parameter :: pc = (/ 2, 10, 20 /)
```
If so, can we add a test for that too?

pearu replied:
yes, the order of attribute specifications is insignificant.

eric-wieser commented on numpy/f2py/tests/test_crackfortran.py:
```diff
@@ -89,6 +89,49 @@ def test_defaultPublic(self, tmp_path):
+                selected_real_kind=crackfortran._selected_real_kind_func)
+        params = {}
+        dimspec = ''
+        assert_raises(ValueError, crackfortran.param_eval, v, g_params, params,
+                      dimspec=dimspec)
```
A similar test for a too-many-dimensional range would be good too.

melissawm commented:
I think I got it, let me know if you had something else in mind. Thanks again!

HaoZeke commented on numpy/f2py/crackfortran.py:
```diff
@@ -2735,19 +2774,23 @@ def analyzevars(block):
+                        [x.strip() for x in markoutercomma(dim).split('@,@')]
+                ):
+                    # d is the expression inside the dimension declaration
                     star = '*'
                     if d == ':':
                         star = ':'
```
I'm not sure I understand the variable naming scheme here; `star = ':'` seems a bit strange.

HaoZeke replied:
It might be nicer to directly assign `dl=[:]` here.

HaoZeke replied:
Also this should be moved below the `try...except` block; since `param_parse` can also return `:`.

melissawm replied:
> It might be nicer to directly assign dl=[:] here.

Because `d` might be different from either `*` or `:`, this wouldn't be as direct. I'll change the name of this variable but I'd prefer leaving the logic as it is for now and maybe fixing it in a follow-up, if that's ok.

HaoZeke reviewed (commented):
Generally looks great; apart from a few minor comments I think it should be good to go. Love the docstrings ^_^.

Also this will close #8730 so the main text should be updated.

melissawm commented on numpy/f2py/crackfortran.py:
```diff
@@ -2567,11 +2590,18 @@ def get_parameters(vars, global_params={}):
                 outmess(f'get_parameters[TODO]: '
                         f'implement evaluation of complex expression {v}')
 
+            dimspec = ([s.lstrip('dimension') for s in vars[n]['attrspec']
+                       if s.startswith('dimension')] or [None])[0]
+
+            dimspec = ([s.lstrip('dimension') for s in vars[n]['attrspec']
+                       if s.startswith('dimension')] or [None])[0]
```
This must have been a rebase gone wrong, thanks for catching it!

melissawm commented:
I'll rebase and fix any conflicts left.

melissawm commented:
(Sorry for the mess in the history - I didn't squash yet because it might make it harder to review. I'll do that once this is ready to merge)

pearu reviewed (dismissed):
Found one bug and have a few nits. Otherwise, looks good!

HaoZeke commented:
@melissawm could you squash this before I rebase it?

melissawm commented:
Done, thanks @HaoZeke

HaoZeke commented:
Thanks @melissawm. This actually rebases cleanly, the relevant branch [is here](https://github.com/HaoZeke/numpy/tree/f2py-param_fix_15457). So:

```bash
# Grab the F2PY refactor PR
gh pr checkout 20481
# Add it to your fork, assumes origin to be the fork
git push origin HEAD:f2pyRefactor
# Goto your branch for this PR
git checkout f2py-param-fix
# Rebase onto the refactor
git rebase -i f2pyRefactor
# Push the updated branch
git push -f
```

In case it does not apply cleanly, the other option is to checkout the branch mentioned above and force push that to this PR. Let me know if this can be made clearer for the other PRs :)

HaoZeke reviewed (approved):
Perhaps a release note for `param_eval` would be nice, otherwise this is ready.

HaoZeke reviewed (changes requested):
This needs some work after the changes introduced in the recent PRs, but the issue still exists.

bilderbuchi commented:
@melissawm what's the status of this PR? It seems it was ~finished/ready to go already?
I just hit #11612, and this PR got my hopes for a fix up :D

charris commented:
@melissawm Is this still in progress?

@pearu @HaoZeke Have your requested changes been dealt with?

charris commented:
Needs rebase.

HaoZeke committed:
TST: Cleanup rebase issue, no prints in tests

HaoZeke reviewed (approved):
This works, and fixes the still present https://github.com/numpy/numpy/issues/11612 so it would be great to have this in :)

I think all the requested changes were made, and subsequent iterations can clean up any new issues.

mattip commented:
@pearu could you take a look?

HaoZeke commented:
Would be nice to have this in :) I think any additional changes can be done in followups if necessary.

HaoZeke commented:
@pearu could you take a look / remove the review changes requested note? Can't be merged otherwise
